src/sqlfluff/core/rules/analysis/select_crawler.py

```diff
@@ -81,7 +81,10 @@ def crawl(
             if seg.type == "table_reference":
                 if not seg.is_qualified() and seg.raw in queries:
                     # It's a CTE.
-                    yield queries[seg.raw]
+                    # :TRICKY: Pop the CTE from "queries" to help callers avoid
+                    # infinite recursion. We could make this behavior optional
+                    # someday, if necessary.
+                    yield queries.pop(seg.raw)
```

Comment:
This is the crux of the fix -- avoids recursion by removing CTEs as they are processed.